### ISSUE TITLE
HTTP 500 errors are not handled correctly

### ISSUE THREAD

**Issue by john-mcnamara-intel** [getpatchwork/git-pw#14]
Hi,

I have an issue running git-pw against the recently upgraded patchwork for the DPDK project: http://dpdk.org/dev/patchwork/project/dpdk/

I installed and configured git-pw as follows:

```bash
pip install git-pw

git-pw --version
   git-pw, version 1.1.2

git config pw.server   http://dpdk.org/dev/patchwork/
git config pw.project  dpdk
git config pw.username johnmcnamara
git config pw.password password
```

Running commands such as `git-pw patch list --state new` work fine.

However I have issues applying a patch:

```
$ git-pw --debug patch apply 40648 
2018-06-06 19:35:01,556 - git_pw.patch - INFO - Applying patch: id=40648, series=None, deps=True, args=
2018-06-06 19:35:01,565 - git_pw.config - DEBUG - Retrieved 'server' setting from git-config
2018-06-06 19:35:01,565 - git_pw.config - DEBUG - Retrieved 'server' setting from cache
2018-06-06 19:35:01,566 - git_pw.api - DEBUG - GET http://dpdk.org/dev/patchwork/api/patches/40648/
2018-06-06 19:35:01,580 - git_pw.config - DEBUG - Retrieved 'username' setting from git-config
2018-06-06 19:35:01,587 - git_pw.config - DEBUG - Retrieved 'password' setting from git-config
2018-06-06 19:35:01,587 - git_pw.config - DEBUG - Retrieved 'username' setting from cache
2018-06-06 19:35:01,587 - git_pw.config - DEBUG - Retrieved 'password' setting from cache
2018-06-06 19:35:01,591 - requests.packages.urllib3.connectionpool - INFO - Starting new HTTP connection (1): proxy.ir.intel.com
2018-06-06 19:35:01,980 - requests.packages.urllib3.connectionpool - DEBUG - "GET http://dpdk.org/dev/patchwork/api/patches/40648/ HTTP/1.1" 200 None
2018-06-06 19:35:02,007 - git_pw.api - DEBUG - Got response
2018-06-06 19:35:02,011 - git_pw.api - DEBUG - GET http://dpdk.org/dev/patchwork/patch/40648/mbox/
2018-06-06 19:35:02,019 - git_pw.config - DEBUG - Retrieved 'username' setting from cache
2018-06-06 19:35:02,019 - git_pw.config - DEBUG - Retrieved 'password' setting from cache
2018-06-06 19:35:02,019 - git_pw.config - DEBUG - Retrieved 'username' setting from cache
2018-06-06 19:35:02,019 - git_pw.config - DEBUG - Retrieved 'password' setting from cache
2018-06-06 19:35:02,022 - requests.packages.urllib3.connectionpool - INFO - Starting new HTTP connection (1): proxy.ir.intel.com
2018-06-06 19:35:02,176 - requests.packages.urllib3.connectionpool - DEBUG - "GET http://dpdk.org/dev/patchwork/patch/40648/mbox/?series=%2A HTTP/1.1" 500 None
Traceback (most recent call last):
  File "/usr/bin/git-pw", line 11, in <module>
    sys.exit(cli())
  File "/usr/lib/python2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/git_pw/patch.py", line 41, in apply_cmd
    mbox = api.download(patch['mbox'], {'series': series})
  File "/usr/lib/python2.7/site-packages/git_pw/api.py", line 141, in download
    rsp = get(url, params, stream=True)
  File "/usr/lib/python2.7/site-packages/git_pw/api.py", line 103, in get
    _handle_error('fetch', exc)
  File "/usr/lib/python2.7/site-packages/git_pw/api.py", line 81, in _handle_error
    LOG.error(exc.response.json())
  File "/usr/lib/python2.7/site-packages/requests/models.py", line 826, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/lib64/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
```
Running http://dpdk.org/dev/patchwork/patch/40648/mbox/?series=%2A  manually fails. 
However, http://dpdk.org/dev/patchwork/patch/40648/mbox/?series= (without the encoded *) works. 

The command also works if I modify git-pw to use `''` instead of `'*'` for undefined series.

Is this a server config issue or a client issue?

John

**Comment by stephenfin** [comment]
Thanks for reporting this, John. This was an oversight on my end. The patch at http://dpdk.org/dev/patchwork/patch/40648/ does not have an associated series as it was processed before the upgrade to Patchwork 2.0 (which added series) support. As a result, attempting to download the patch with it's dependencies fails. Clearly we're missing a check to ensure a patch has a series before we attempt to build the dependencies.

There are two fixes necessary here. Firstly, a server-side fix to ensure a patch has a series. If this is not the case, we should default to download the patch without dependencies. Secondly, `git-pw` should handle HTTP 500 errors better than it does currently. I'm going to retarget this bug at the latter and open a new bug against getpatchwork/patchwork for the former.

**Comment by stephenfin** [comment]
This was resolved in a573972f32afd3b025c14eb194d16a47a347b8b2. Not sure why the linking didn't happen.
